Simple composed sender compiles with thread_scheduler but not time_scheduler

I made a simple example to try out the library, consisting of two `just_result_of`'s followed by a `when_all` and a `then`. It compiles and runs when `thread_scheduler` is used but fails to compile with `time_scheduler`.

The compile error is large and complex but in the end comes down to a tuple index out of range, which I didn't know how to debug.

Reduced example (will compile but not run if fixed because I left out the timer implementation):

```c++
#include <async/schedulers/thread_scheduler.hpp>
#include <async/schedulers/time_scheduler.hpp>
#include <async/schedulers/timer_manager.hpp>
#include <async/just.hpp>
#include <async/just_result_of.hpp>
#include <async/start_detached.hpp>
#include <async/start_on.hpp>
#include <async/sync_wait.hpp>
#include <async/then.hpp>
#include <async/when_all.hpp>

#include <chrono>
#include <iostream>

//
// BOILERPLATE FROM THE DOCS
// for timer_scheduler. Will not link, just for compilation
//

// provide a HAL for the timer scheduler

namespace {
// A HAL defines a time_point type and a task type,
// and provides functions to control a timer interrupt
struct asio_hal {
    using time_point_t = std::chrono::steady_clock::time_point;
    using task_t = async::timer_task<time_point_t>;

    static auto enable() -> void;
    static auto disable() -> void;
    static auto set_event_time(time_point_t tp) -> void;
    static auto now() -> time_point_t;
};

// use the generic timer manager
using timer_manager_t = async::generic_timer_manager<asio_hal>;
} // namespace

// tell the library how to infer a time point type from a duration type by
// specializing time_point_for
template <typename Rep, typename Period>
struct async::timer_mgr::time_point_for<std::chrono::duration<Rep, Period>> {
    using type = asio_hal::time_point_t;
};

// time_scheduler will use this timer_manager
template <> inline auto async::injected_timer_manager<> = timer_manager_t{};

int main() {
    using namespace std::literals;
    using namespace async;

    auto s0 = start_on(time_scheduler{4s},
//    auto s0 = start_on(thread_scheduler{},   // compiles
                       async::just_result_of([](){ return 42; }));
    auto s1 = start_on(time_scheduler{6s},
//    auto s1 = start_on(thread_scheduler{},   // compiles
                       async::just_result_of([](){ return 13; }));
    auto combined = then(when_all(s0, s1), [](int a, int b){ return a*b; });
    
    auto result = sync_wait(combined).value();   // 42*13, delivered shortly after 6s when s1 completes

    std::cout << stdx::get<0>(result) << "\n";

}

```

Thanks for the quick response! It does look like my original testcase is failing, but for a different reason this time:

```
include/async/env.hpp:56:33: error: member ‘stdx::v1::detail::element<1, const async::env<>&>::value’ is uninitialized reference
```

--- a/include/async/start_detached.hpp
+++ b/include/async/start_detached.hpp
@@ -60,7 +60,7 @@ template <typename StopSource, typename Env>
 struct op_state_base {
     using env_t = env<prop<get_stop_token_t,
                            decltype(std::declval<StopSource>().get_token())>,
-                      Env const &>;
+                      Env>;
 
     constexpr explicit(true) op_state_base(Env &&env) : e{std::move(env)} {}
     constexpr op_state_base(op_state_base &&) = delete;
@@ -68,8 +68,7 @@ struct op_state_base {
     template <stdx::ct_string> auto die() {}
 
     [[nodiscard]] constexpr auto query(get_env_t) const -> env_t {
-        return env{prop{get_stop_token_t{}, stop_src.get_token()},
-                   std::cref(e)};
+        return env{prop{get_stop_token_t{}, stop_src.get_token()}, e};
     }
 
     using stop_source_t = StopSource;

--- a/include/async/sync_wait.hpp
+++ b/include/async/sync_wait.hpp
@@ -60,7 +60,7 @@ using sync_wait_type = value_types_of_t<S, E, decayed_tuple, std::optional>;
 template <typename Uniq, sender S, typename Env> auto wait(S &&s, Env &&e) {
     run_loop<Uniq> rl{};
     auto sched = rl.get_scheduler();
-    auto new_env = env{prop{get_scheduler_t{}, sched}, std::cref(e)};
+    auto new_env = env{prop{get_scheduler_t{}, sched}, std::forward<Env>(e)};
 
     using E = decltype(new_env);
     using V = detail::sync_wait_type<E, S>;

--- a/include/async/when_all.hpp
+++ b/include/async/when_all.hpp
@@ -448,7 +448,9 @@ template <stdx::ct_string Name, typename... Sndrs> struct sender : Sndrs... {
     template <typename R>
         requires(
             ... and
-            multishot_sender<typename Sndrs::sender_t, std::remove_cvref_t<R>>)
+            multishot_sender<
+                typename Sndrs::sender_t,
+                detail::universal_receiver<env_of_t<std::remove_cvref_t<R>>>>)
     [[nodiscard]] constexpr auto connect(
         R &&r) const & -> op_state_t<Name, std::remove_cvref_t<R>, Sndrs...> {
         check_connect<sender const &, R>();

--- a/include/async/when_any.hpp
+++ b/include/async/when_any.hpp
@@ -420,7 +420,9 @@ struct sender : Sndrs... {
     template <typename R>
         requires(
             ... and
-            multishot_sender<typename Sndrs::sender_t, std::remove_cvref_t<R>>)
+            multishot_sender<
+                typename Sndrs::sender_t,
+                detail::universal_receiver<env_of_t<std::remove_cvref_t<R>>>>)
     [[nodiscard]] constexpr auto connect(R &&r) const
         & -> op_state_t<Name, StopPolicy, std::remove_cvref_t<R>, Sndrs...> {
         check_connect<sender const &, R>();

--- a/test/schedulers/time_scheduler.cpp
+++ b/test/schedulers/time_scheduler.cpp
@@ -1,10 +1,13 @@
 #include "detail/common.hpp"
 
 #include <async/debug.hpp>
+#include <async/just.hpp>
 #include <async/just_result_of.hpp>
 #include <async/schedulers/time_scheduler.hpp>
 #include <async/schedulers/timer_manager.hpp>
 #include <async/start_on.hpp>
+#include <async/then.hpp>
+#include <async/when_all.hpp>
 
 #include <stdx/concepts.hpp>
 #include <stdx/ct_format.hpp>
@@ -326,4 +329,18 @@ TEST_CASE("time_scheduler with no argument is cancellable",
     CHECK(not enabled<default_domain>);
     CHECK(var == 17);
     CHECK(async::timer_mgr::is_idle());
+}
+
+TEST_CASE("multishot connection with when_all", "[time_scheduler]") {
+    using namespace std::chrono_literals;
+    int value{};
+    auto s1 = async::just(42);
+    auto s2 = async::time_scheduler{1s}.schedule();
+    auto w = async::when_all(s1, s2);
+    auto t = w | async::then([&](int x) { value = x; });
+
+    auto op = async::connect(t, receiver{[] {}});
+    async::start(op);
+    async::timer_mgr::service_task();
+    CHECK(value == 42);
 }

